### ISSUE TITLE
porting no-inferrable-return-type to tslint-consistent-codestyle

### ISSUE THREAD

**Issue by ajafff** [ibezkrovnyi/tslint-rules#2]
Continuing discussion from https://github.com/ibezkrovnyi/tslint-rules/issues/1#issuecomment-397099093

> Are you interested/can we somehow integrate no-inferred-return-types (or choose different name you like) into tslint-consistent-codestyle, so all related rules will be in one single repo?

I'm definitely not opposed to that idea. I'd like to know why `no-unnecessary-type-annotation` doesn't work for you? Does it check too much for your use case?

---

Note that I'm eventually moving away from developing TSLint rules in favour of my own project [Fimbullinter](https://github.com/fimbullinter/wotan). Since I'm using `tslint-consistent-codestyle` in all of my projects and even at work, abandoning it is not an option. Therefore I will still maintain that package until the functionality is ported to my new linter engine (these are then made available for TSLint users as well).

**Comment by ibezkrovnyi** [comment]
no-unnecessary-type-annotation is good, but it looks like it doesn't support checking return types

## src/file1.ts
```ts
function a(): number {
  return 5 as number;
}

function b(): number {
  return a();
}
```

## tsling.json
```json5
{
  "extends": [
    "tslint:recommended",
    "tslint-consistent-codestyle",
    "@ibezkrovnyi/tslint-rules"
  ],
  "rules": {
    "no-inferrable-return-types": true,
    "no-unnecessary-type-annotation": [true, "check-return-type"],
    // "no-unnecessary-type-annotation": { "severity": "error", "options": { "check-return-type": true } },
    "no-inferrable-types": true
  }
}
```

## tslint output
```bash
D:\projects\my\typescript\no-unnecessary-type-annotation>npm run lint

> no-unnecessary-type-annotation@1.0.0 lint D:\projects\my\typescript\no-unnecessary-type-annotation
> tslint --format verbose --config tslint.json --project tsconfig.json

ERROR: (no-inferrable-return-types) D:/projects/my/typescript/no-unnecessary-type-annotation/src/file1.ts[1, 13]: type annotation is redundant
ERROR: (no-inferrable-return-types) D:/projects/my/typescript/no-unnecessary-type-annotation/src/file1.ts[5, 13]: type annotation is redundant
```
[project-source-to-reproduce.zip](https://github.com/ibezkrovnyi/tslint-rules/files/2103633/no-unnecessary-type-annotation.zip)

**Comment by ajafff** [comment]
@ibezkrovnyi You are absolutely right. `no-unnecessary-type-annotation` only complains about return type annotations if there is a contextual type for the function.
`no-inferrable-return-types` on the other hand checks if the return type annotation is equal to the type of each `return`.

I think this rule would be a great addition.
Looking at the implementation I can see there is a really nasty hack to trick the type checker. Maybe that can be implemented in a different way. Though I don't consider that a blocker.

**Comment by ibezkrovnyi** [comment]
> I think this rule would be a great addition.

again, It would be great to have everything in one repo :clap:

[Here I described function names](https://github.com/palantir/tslint/issues/695#issuecomment-367192971) which may help to implement this rule (and other rules from your lib and tslint itself) without hacks and workarounds, but I don't want to copy code from typescript source, because internal logic may  and will change... And small trick looks here less evil for now.
